### PR TITLE
Added similarity of documents and tokens

### PR DESCRIPTION
**Pull request by bicycleman15**
## Description

I just thought it would be a nice addition to include similarity between tokens.
I have added similarity between docs as well on the basis of averaged token vectors.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

```python
doc1 = nlp("hello banana")
doc2 = nlp("artic monkeys")
token1 = doc1[0]
token2 = doc1[1]

token1.vector_norm 
# L2 norm

token1.similarity(token2)
# cosine similarity

doc1.vector_norm
# L2 norm of doc

doc1.similarity(doc2)
# cosine similarity
```

## Checklist:

* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
* [x] I have added tests for my changes

### CODE REVIEW

**Comment by bicycleman15** [comment]
Why aren't the CI tests passing ? They are passing on my machine. Any help @AlanAboudib @sachin-101 ??

**Comment by sachin-101** [comment]
The tests indicate that black tests are failing. Is your branch updated with latest master? Cause we have black support. Before any commit it shows if any of the files are failing black test.

If not that then it may be due to `E   OSError: [Errno 12] Cannot allocate memory`

**Comment by bicycleman15** [comment]
Yeah I saw that. Any way to fix this cannot allocate memory issue ?

File: syfertext/doc.py [hunk 1]
```diff
@@ -155,6 +155,37 @@ def vector(self):
...
+        """
+        Get L2 norm of document vector
+
+        Returns:
+            Tensor(float): L2 norm of the document vector
+        """
+        vector = torch.tensor(self.vector)
+        total = (vector ** 2).sum()
+
+        return torch.sqrt(total)
```
Comment (AlanAboudib) [inline syfertext/doc.py:169]:
```suggestion
    def vector_norm(self) -> torch.FloatTensor:
        """Get the L2 norm of the document vector

        Returns:
            A torch tensor representing the L2 norm of the document vector
        """
        
        vector = torch.tensor(self.vector)
        
        norm = (vector ** 2).sum()
        norm = torch.sqrt(norm)

        return norm
```

File: syfertext/doc.py [hunk 2]
```diff
@@ -155,6 +155,37 @@ def vector(self):
...
+        """
+
+        assert (
+            self.vector_norm.item() != 0.0 and other.vector_norm.item() != 0.0
+        ), "one or both of the tokens is invalid"
+
+        return torch.dot(torch.tensor(self.vector), torch.tensor(other.vector)) / (
+            self.vector_norm * other.vector_norm
+        )
+
```
Comment (AlanAboudib) [inline syfertext/doc.py:188]:
```suggestion
    def similarity(self, other: "Doc") -> torch.Tensor:
        """Compute the cosine similarity between two Doc vectors.
        
        Args:
            other (Doc): The Doc to compare with.
        
        Returns:
            Tensor: A cosine similarity score. Higher is more similar.
        """

        # Make sure both vectors have non-zero norms
        assert (
            self.vector_norm.item() != 0.0 and other.vector_norm.item() != 0.0
        ), "One of the provided vectors has a zero norm!"

        # Compute similarity
        sim = torch.dot(torch.tensor(self.vector), torch.tensor(other.vector))
        sim /= self.vector_norm * other.vector_norm

        return  sim
```

File: syfertext/token.py [hunk 3]
```diff
@@ -53,6 +53,37 @@ def vector(self):
...
+
+        Returns: 
+            Tensor: The L2 norm of the vector representation.
+        """
+
+        vector = torch.tensor(self.vector)
+        total = (vector ** 2).sum()
+
+        return torch.sqrt(total)
+
```
Comment (AlanAboudib) [inline syfertext/token.py:68]:
```suggestion
    def vector_norm(self) -> torch.Tensor:
        """The L2 norm of the token's vector representation.

        Returns: 
            Tensor: The L2 norm of the vector representation.
        """
        
        # Convert the vector from a numpy array to a Tensor
        vector = torch.tensor(self.vector)
        
        # Compute the norm
        norm = (vector ** 2).sum()
        norm = torch.sqrt(norm)

        return norm
```

File: syfertext/token.py [hunk 4]
```diff
@@ -53,6 +53,37 @@ def vector(self):
...
+        """
+
+        assert (
+            self.vector_norm.item() != 0.0 and other.vector_norm.item() != 0.0
+        ), "one of the token is invalid"
+
+        return torch.dot(torch.tensor(self.vector), torch.tensor(other.vector)) / (
+            self.vector_norm * other.vector_norm
+        )
+
```
Comment (AlanAboudib) [inline syfertext/token.py:86]:
```suggestion
    def similarity(self, other):
        """Compute the cosine similarity between tokens' vectors.
        
        Args:
            other (Token): The Token to compare with.
        
        Returns:
            Tensor: A cosine similarity score. Higher is more similar.
        """
        
        # Make sure both vectors have non-zero norms
        assert (
            self.vector_norm.item() != 0.0 and other.vector_norm.item() != 0.0
        ), "One of the provided tokens has a zero norm."

        # Compute similarity
        sim = torch.dot(torch.tensor(self.vector), torch.tensor(other.vector))
        sim /= self.vector_norm * other.vector_norm

        return  sim

```

File: tests/local_mode/test_token.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    """Test that the similarity of valid tokens"""
+    doc = nlp("hello banana")
+    token1 = doc[0]
+    token2 = doc[1]
+
+    # check commutativity of similarity
+    assert token1.similarity(token2) == token2.similarity(token1)
+
+    # check if similarity is in valid range
+    assert -1.0 <= token1.similarity(token2).item() <= 1.0
```
Comment (AlanAboudib) [inline tests/local_mode/test_token.py:40]:
```suggestion
def test_valid_token_norm_is_not_zero():
    """Test that the norm of a valid token is not zero"""
    
    doc = nlp("banana")
    norm = doc[0].vector_norm.item()

    # check that norm is not zero for valid token
    assert norm != 0.0


def test_oov_token_norm_is_zero():
    """Test that the norm of an out-of-vocab oken is zero"""
    
    doc = nlp("notvalidtoken")
    norm = doc[0].vector_norm.item()

    # check that norm is zero for invalid token
    assert norm == 0.0


def test_similarity_tokens():
    """Test that the similarity of valid tokens"""
    
    doc = nlp("hello banana")
    token1 = doc[0]
    token2 = doc[1]

    # check commutativity of similarity
    assert token1.similarity(token2) == token2.similarity(token1)

    # check if similarity is in valid range
    assert -1.0 <= token1.similarity(token2).item() <= 1.0
```

**Comment by AlanAboudib** [comment]
@bicycleman15 Just needs blacking

**Comment by bicycleman15** [comment]
Done 👍

**Pull request merged**